src/golang/lib/workflow/operator/operator.go

```diff
@@ -0,0 +1,134 @@
+package operator
+
+import (
+	"context"
+	"github.com/aqueducthq/aqueduct/lib/collections/operator"
+	"github.com/aqueducthq/aqueduct/lib/collections/operator_result"
+	"github.com/aqueducthq/aqueduct/lib/collections/shared"
+	"github.com/aqueducthq/aqueduct/lib/database"
+	"github.com/aqueducthq/aqueduct/lib/job"
+	"github.com/aqueducthq/aqueduct/lib/vault"
+	"github.com/aqueducthq/aqueduct/lib/workflow/artifact"
+	"github.com/dropbox/godropbox/errors"
+	"github.com/google/uuid"
+)
+
+// Operator is an interface for managing and inspecting the lifecycle of an operator
+// used by a workflow run.
+type Operator interface {
+	Type() operator.Type
+	Name() string
+	ID() uuid.UUID
+	JobSpec() job.Spec
+
+	// Ready indicates whether this operator is ready to be scheduled. That is to say,
+	// all upstream dependencies have been successfully computed.
+	Ready(ctx context.Context) bool
+
+	// GetExecState performs a non-blocking fetch for the execution state of this operator.
+	GetExecState(ctx context.Context) (*shared.ExecutionState, error)
+
+	// PersistResult writes the results of this operator execution to the database.
+	// This method also persists any artifact results produced by this operator.
+	PersistResult(ctx context.Context) error
+
+	// Finish is an end-of-lifecycle hook meant to do any final cleanup work.
+	// Recursively calls Finish() on all the operator's output artifacts.
+	Finish(ctx context.Context)
+}
+
+func initializeOperatorResultInDatabase(
+	ctx context.Context,
+	opID uuid.UUID,
+	workflowDagResultID uuid.UUID,
+	opResultWriter operator_result.Writer,
+	db database.Database,
+) (uuid.UUID, error) {
+	operatorResult, err := opResultWriter.CreateOperatorResult(
+		ctx,
+		workflowDagResultID,
+		opID,
+		db,
+	)
+	if err != nil {
+		return uuid.Nil, errors.Wrap(err, "Failed to create operator result record.")
+	}
+	return operatorResult.Id, nil
+}
+
+func NewOperator(
+	ctx context.Context,
+	dbOperator operator.DBOperator,
+	inputs []artifact.Artifact,
+	inputContentPaths []string,
```

Comment:
The interface was specifically designed to avoid having to deal with path info. When only need the paths when initializing an instance of this interface.

Reply:
I see, yeah I wish go support things like `[]artifact.BaseArtifact` and we could use anything that inherits `baseArtifact` here. But now I agree the current way is fine.

We could put all paths in a struct though like `struct{ContentPath str, MetadataPath str}` and reuse in a few places, but it's less a big deal given we only have two for now